### PR TITLE
WIP FileSystem strong typing: FileUrl

### PR DESCRIPTION
**Pull request by chibenwa**
So much knowledge encoding in raw string is not viable.

TODO:

 - FileUrlTest
 - Fix the very likely failing test cases

### CODE REVIEW

**Comment by trantienduchn** [comment]
> why defining a FileURL, and not using directly File?

FileUrl is close to James File Schema. maybe.

**Comment by chibenwa** [comment]
Several reasons:
 - FileUrl unifies File and classpath + relative and absolute files. We need to encapsulate the knowledge of these URL
 - `File getFile(File xx)` would be a silly method, isn't it?
 - Using POJO in configuration felt better than File that requires a Filesystem at hands.

File: server/container/filesystem-api/src/main/java/org/apache/james/filesystem/api/FileUrl.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+package org.apache.james.filesystem.api;
+
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+
+import com.google.common.base.Preconditions;
+
+public class FileUrl {
```
Comment (mbaechler) [inline server/container/filesystem-api/src/main/java/org/apache/james/filesystem/api/FileUrl.java]: When you propose such an important concept, I think it makes sense to write an extensive javadoc to explain what are the design choices you propose.

**Comment by chibenwa** [comment]
This work is non conclusive

**Pull request closed without merging**